File: app/Models/ServiceRequest.php
<?php

namespace App\Models;

use App\Enums\ServiceRequestStatus;
use App\Enums\ServiceUrgency;
use App\Policies\ServiceRequestPolicy;
use Database\Factories\ServiceRequestFactory;
use Illuminate\Database\Eloquent\Attributes\Fillable;
use Illuminate\Database\Eloquent\Attributes\UsePolicy;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\Relations\HasOne;

#[Fillable(['service_finder_id', 'requested_provider_id', 'service_id', 'province_id', 'municipality_id', 'title', 'description', 'urgency', 'needed_at', 'budget_min', 'budget_max', 'status', 'safety_acknowledged_at'])]
#[UsePolicy(ServiceRequestPolicy::class)]
class ServiceRequest extends Model
{
    /** @use HasFactory<ServiceRequestFactory> */
    use HasFactory;

    protected function casts(): array
    {
        return ['urgency' => ServiceUrgency::class, 'needed_at' => 'datetime', 'budget_min' => 'decimal:2', 'budget_max' => 'decimal:2', 'status' => ServiceRequestStatus::class, 'safety_acknowledged_at' => 'datetime'];
    }

    public function serviceFinder(): BelongsTo
    {
        return $this->belongsTo(User::class, 'service_finder_id');
    }

    public function requestedProvider(): BelongsTo
    {
        return $this->belongsTo(User::class, 'requested_provider_id');
    }

    public function service(): BelongsTo
    {
        return $this->belongsTo(Service::class);
    }

    public function province(): BelongsTo
    {
        return $this->belongsTo(Province::class);
    }

    public function municipality(): BelongsTo
    {
        return $this->belongsTo(Municipality::class);
    }

    public function job(): HasOne
    {
        return $this->hasOne(Job::class);
    }
}

File: mobile/oncall_mobile/lib/models/service_request.dart
import '../core/json.dart';

class ServiceRequest {
  ServiceRequest({
    required this.id,
    required this.title,
    this.description,
    required this.urgency,
    this.neededAt,
    this.budgetMin,
    this.budgetMax,
    required this.status,
    this.service,
    this.province,
    this.municipality,
    this.serviceFinder,
    this.requestedProvider,
    this.jobId,
    this.createdAt,
  });

  factory ServiceRequest.fromJson(Map<String, dynamic> json) => ServiceRequest(
    id: json['id'] as int,
    title: json['title'] as String,
    description: json['description'] as String?,
    urgency: json['urgency'] as String,
    neededAt: parseDate(json['needed_at']),
    budgetMin: parseNum(json['budget_min']),
    budgetMax: parseNum(json['budget_max']),
    status: json['status'] as String,
    service: IdName.fromJsonOrNull(json['service']),
    province: IdName.fromJsonOrNull(json['province']),
    municipality: IdName.fromJsonOrNull(json['municipality']),
    serviceFinder: IdName.fromJsonOrNull(json['service_finder']),
    requestedProvider: IdName.fromJsonOrNull(json['requested_provider']),
    jobId: json['job_id'] as int?,
    createdAt: parseDate(json['created_at']),
  );

  final int id;
  final String title;
  final String? description;
  final String urgency;
  final DateTime? neededAt;
  final double? budgetMin;
  final double? budgetMax;
  final String status;
  final IdName? service;
  final IdName? province;
  final IdName? municipality;
  final IdName? serviceFinder;
  final IdName? requestedProvider;
  final int? jobId;
  final DateTime? createdAt;

  bool get isOpen => status == 'REQUESTED' || status == 'SEARCHING';
}
